pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -421,30 +367,101 @@ func (s *NodeController) checkNodeReady(node *api.Node) *api.NodeCondition {
 	}
 }
 
-// deletePods will delete all pods from master running on given node.
-func (s *NodeController) deletePods(nodeID string) error {
-	glog.V(2).Infof("Delete all pods from %v", nodeID)
-	// TODO: We don't yet have field selectors from client, see issue #1362.
-	pods, err := s.kubeClient.Pods(api.NamespaceAll).List(labels.Everything())
+// PopulateAddresses queries Address for given list of nodes.
+func (s *NodeController) PopulateAddresses(nodes *api.NodeList) (*api.NodeList, error) {
+	if s.isRunningCloudProvider() {
+		instances, ok := s.cloud.Instances()
+		if !ok {
+			return nodes, ErrCloudInstance
+		}
+		for i := range nodes.Items {
+			node := &nodes.Items[i]
+			nodeAddresses, err := instances.NodeAddresses(node.Name)
+			if err != nil {
+				glog.Errorf("error getting instance addresses for %s: %v", node.Name, err)
+			} else {
+				node.Status.Addresses = nodeAddresses
+			}
+		}
+	} else {
+		for i := range nodes.Items {
+			node := &nodes.Items[i]
+			addr := net.ParseIP(node.Name)
+			if addr != nil {
+				address := api.NodeAddress{Type: api.NodeLegacyHostIP, Address: addr.String()}
+				node.Status.Addresses = []api.NodeAddress{address}
+			} else {
+				addrs, err := s.lookupIP(node.Name)
+				if err != nil {
+					glog.Errorf("Can't get ip address of node %s: %v", node.Name, err)
+				} else if len(addrs) == 0 {
+					glog.Errorf("No ip address for node %v", node.Name)
+				} else {
+					address := api.NodeAddress{Type: api.NodeLegacyHostIP, Address: addrs[0].String()}
+					node.Status.Addresses = []api.NodeAddress{address}
+				}
+			}
+		}
+	}
+	return nodes, nil
+}
+
+// MonitorNodeStatus verifies node status are constantly updated by kubelet, and if
+// not, post node NotReady status. It also evicts all pods if node is not ready for
+// a long period of time.
+func (s *NodeController) MonitorNodeStatus() error {
+	nodes, err := s.kubeClient.Nodes().List()
 	if err != nil {
 		return err
 	}
-	for _, pod := range pods.Items {
-		if pod.Status.Host != nodeID {
-			continue
+	for i := range nodes.Items {
+		node := &nodes.Items[i]
+		// Precompute all condition times to avoid deep copy of node status (We'll modify node for
+		// updating, and NodeStatus.Conditions is an array, which makes assignment copy not useful).
+		latestConditionTime := s.latestConditionTime(node, api.NodeReady)
+		var gracePeriod time.Duration
+		if latestConditionTime == node.CreationTimestamp {
+			gracePeriod = nodeStartupGracePeriod
+		} else {
+			gracePeriod = nodeMonitorGracePeriod
 		}
-		glog.V(2).Infof("Delete pod %v", pod.Name)
-		if err := s.kubeClient.Pods(pod.Namespace).Delete(pod.Name); err != nil {
-			glog.Errorf("Error deleting pod %v: %v", pod.Name, err)
+		latestFullConditionTime := s.latestConditionTimeWithStatus(node, api.NodeReady, api.ConditionFull)
```

Comment:
remove this line and inline the definition of this variable where it is used at the end of the function, since that's the only place where it's used

also, please update the comment to "NodeReady condition was last set longer ago than gracePeriod, so update it to ConditionNone (regardless of its current value) in the master, without contacting kubelet" -- I think that's clearer

also, I didn't really understand the logic here. IIUC the condition you're testing, now > latestConditionTime + gracePeriod, means the last time nodeReady condition was set for the node, regardless of whether it was set to full or none?